tests/core/update-snapd-symlink/task.yaml

```diff
@@ -0,0 +1,93 @@
+summary: Verify that snapd services are restarted with the right symlink
+
+details: |
+  When updating the snapd snap, we need to restart some
+  services. Those services need to be restarted when the symlink is
+  updated. And if the services fail, then we need to reset the
+  symlink.
+
+systems:
+  - ubuntu-core-20-*
+  - ubuntu-core-22-*
+
+restore: |
+  current_rev="$(readlink /snap/snapd/current)"
+  rm -rf ./snapd-modified
+  rm -rf ./snapd-broken
+  if [ -r initial-rev ]; then
+    initial_rev="$(cat initial-rev)"
+    rm initial-rev
+    for revno_path in /snap/snapd/*; do
+      revno="$(basename "${revno_path}")"
+      if [ "${revno}" != current ] && [ "${revno}" != "${initial_rev}" ]; then
+        if [ "${revno}" = "${current_rev}" ]; then
+          snap revert snapd --revision="${initial_rev}" || true
+        fi
+        snap remove snapd --revision="${revno}" || true
+      fi
+    done
+  fi
+
+execute: |
+  old_current="$(readlink /snap/snapd/current)"
+  echo "${old_current}" >initial-rev
+
+  unsquashfs -d ./snapd-modified "/var/lib/snapd/snaps/snapd_${old_current}.snap"
+  for binary in snapd snapd-apparmor; do
+    mv "./snapd-modified/usr/lib/snapd/${binary}" "./snapd-modified/usr/lib/snapd/${binary}.real"
+  cat <<\EOF >"./snapd-modified/usr/lib/snapd/${binary}"
+  #!/bin/bash
+  set -eux
+  binary="$(basename "$0")"
+  dir="$(dirname "$0")"
+  realpath "${dir}/${binary}.real" || true
+  realpath "/snap/snapd/current/usr/lib/snapd/${binary}.real" || true
+  stat "${dir}/${binary}.real" || true
+  stat "/snap/snapd/current/usr/lib/snapd/${binary}.real" || true
+  if ! [ "${dir}/${binary}.real" -ef "/snap/snapd/current/usr/lib/snapd/${binary}.real" ]; then
+    echo "Trying to execute the binary '${binary}' before it is available at the the right place" 1>&2
+    exit 1
+  fi
+  exec "${dir}/${binary}.real" "$@"
+  EOF
+    chmod +x "./snapd-modified/usr/lib/snapd/${binary}"
+  done
+  (cd snapd-modified && snap pack .)
+
+  unsquashfs -d ./snapd-broken "/var/lib/snapd/snaps/snapd_${old_current}.snap"
+  rm "./snapd-broken/usr/lib/snapd/snapd"
+  cat <<\EOF >"./snapd-broken/usr/lib/snapd/snapd"
+  #!/bin/bash
+  set -eu
+  echo "INJECT FAILURE" 1>&2
```

Comment:
btw. there's the faultinject framework in snapd, where in the code you can call `MaybeInjectFault("<some-tag>")`, and set `SNAPD_FAULT_INJECT=<some-tag>:panic` in the environment. No need to use it here, but it may come in handy later.